src/state.rs

```diff
@@ -259,6 +267,227 @@ impl DemandProxyState {
+            }
+            let r = resolver_result.unwrap();
+
+            let resp = r.lookup_ip(&hostname).await;
+            if resp.is_err() {
+                warn!(
+                    "system dns async resolution: error response for workload {} is: {:?}",
+                    &workload_uid, resp
+                );
+                return;
+            } else {
+                trace!(
+                    "system dns async resolution: response for workload {} is: {:?}",
+                    &workload_uid,
+                    resp
+                );
+            }
+            let resp = resp.unwrap();
+            let mut dns_refresh_rate = std::time::Duration::from_secs(u64::MAX);
+            let ips = HashSet::from_iter(resp.as_lookup().record_iter().filter_map(|record| {
+                if record.rr_type().is_ip_addr() {
+                    let record_ttl = u64::from(record.ttl());
+                    if let Some(ipv4) = record.data().unwrap().as_a() {
+                        if record_ttl < dns_refresh_rate.as_secs() {
+                            dns_refresh_rate = std::time::Duration::from_secs(record_ttl);
+                        }
+                        return Some(IpAddr::V4(*ipv4));
+                    }
+                    if let Some(ipv6) = record.data().unwrap().as_aaaa() {
+                        if record_ttl < dns_refresh_rate.as_secs() {
+                            dns_refresh_rate = std::time::Duration::from_secs(record_ttl);
+                        }
+                        return Some(IpAddr::V6(*ipv6));
+                    }
+                    return None;
+                }
+                None
+            }));
+            if ips.is_empty() {
+                // if we have no DNS records with a TTL to lean on; lets try to refresh again in 60s
+                dns_refresh_rate = std::time::Duration::from_secs(60);
+            }
+            let now = std::time::Instant::now();
+            let rdns = ResolvedDns {
+                hostname: hostname.to_owned(),
+                ips,
+                initial_query: Some(now),
+                dns_refresh_rate,
+            };
+            state.set_ips_for_hostname(hostname, rdns);
+        })
+    }
+
+    pub fn set_ips_for_hostname(&mut self, hostname: String, rdns: ResolvedDns) {
+        self.state
+            .write()
+            .unwrap()
+            .resolved_dns
+            .by_hostname
+            .insert(hostname, rdns);
+    }
+
+    pub fn get_ips_for_hostname(&mut self, hostname: &String) -> Option<ResolvedDns> {
+        self.state
+            .read()
+            .unwrap()
+            .resolved_dns
+            .by_hostname
+            .get(hostname)
+            .filter(|rdns| {
+                rdns.initial_query.is_some()
+                    && rdns.initial_query.unwrap().elapsed() < rdns.dns_refresh_rate
+            })
+            .cloned()
+    }
+
+    pub async fn set_gateway_address(
```

Comment:
Why did this become a method if there is no reference to `self`?

Reply:
Seems like maybe it would be a method of `Upstream`.

Reply:
great call, got lost in refactor and cleanup. addressed in https://github.com/istio/ztunnel/pull/612/commits/61189f87f24444cfc266f86f35e7c8e5df29c252

Reply:
i left it where it originally lived (check the new PR diff) but im happy to move it